File: nodejs/lib/quote.js
/** GoogleSQL literal quoting and escaping. */

export const QuoteStrategy = Object.freeze({
  LEGACY: 0,
  ALWAYS: 1,
  MIN_ESCAPE: 2,
});

export const PreferredQuote = Object.freeze({
  DOUBLE: 0,
  SINGLE: 1,
});

/**
 * @typedef {Object} LiteralQuoteConfig
 * @property {number} [strategy]
 * @property {number} [preferredQuote]
 */

/** @type {LiteralQuoteConfig} */
export const DEFAULT_LITERAL_QUOTE = Object.freeze({
  strategy: QuoteStrategy.LEGACY,
  preferredQuote: PreferredQuote.DOUBLE,
});

/** @param {LiteralQuoteConfig | null | undefined} cfg */
export function normalizeLiteralQuote(cfg) {
  const input = cfg ?? DEFAULT_LITERAL_QUOTE;
  let strategy = input.strategy ?? QuoteStrategy.LEGACY;
  if (strategy !== QuoteStrategy.LEGACY &&
      strategy !== QuoteStrategy.ALWAYS &&
      strategy !== QuoteStrategy.MIN_ESCAPE) {
    strategy = QuoteStrategy.LEGACY;
  }
  let preferredQuote = input.preferredQuote ?? PreferredQuote.DOUBLE;
  if (preferredQuote !== PreferredQuote.DOUBLE &&
      preferredQuote !== PreferredQuote.SINGLE) {
    preferredQuote = PreferredQuote.DOUBLE;
  }
  return { strategy, preferredQuote };
}

const PRINTABLE_RE = /^[\p{L}\p{M}\p{N}\p{P}\p{S} ]$/u;

/**
 * @param {number} r
 */
function isPrintable(r) {
  if (r === 0x20) {
    return true;
  }
  return PRINTABLE_RE.test(String.fromCodePoint(r));
}

/**
 * @param {string | Uint8Array} payload
 * @param {LiteralQuoteConfig} cfg
 */
export function quoteCharForPayload(payload, cfg) {
  const normalized = normalizeLiteralQuote(cfg);
  /** @type {Uint8Array} */
  let data;
  if (typeof payload === 'string') {
    data = new TextEncoder().encode(payload);
  } else {
    data = payload;
  }

  if (normalized.strategy === QuoteStrategy.ALWAYS) {
    return normalized.preferredQuote === PreferredQuote.SINGLE ? "'" : '"';
  }

  const pref = normalized.preferredQuote === PreferredQuote.SINGLE ? 39 : 34;
  const other = pref === 39 ? 34 : 39;

  if (normalized.strategy === QuoteStrategy.MIN_ESCAPE) {
    let singleCount = 0;
    let doubleCount = 0;
    for (const b of data) {
      if (b === 39) {
        singleCount++;
      } else if (b === 34) {
        doubleCount++;
      }
    }
    if (singleCount < doubleCount) {
      return "'";
    }
    if (doubleCount < singleCount) {
      return '"';
    }
    return String.fromCharCode(pref);
  }

  let hasPref = false;
  for (const b of data) {
    if (b === other) {
      return String.fromCharCode(pref);
    }
    if (b === pref) {
      hasPref = true;
    }
  }
  if (hasPref) {
    return String.fromCharCode(other);
  }
  return String.fromCharCode(pref);
}

/**
 * @param {number} r
 * @param {boolean} isString
 * @param {string} quote
 */
export function escapeRune(r, isString, quote) {
  const q = quote ? quote.charCodeAt(0) : -1;
  if (r === 92 || (quote && r === q)) {
    return '\\' + String.fromCharCode(r);
  }
  if (isString && r === 10) {
    return '\\n';
  }
  if (isString && r === 13) {
    return '\\r';
  }
  if (isString && r === 9) {
    return '\\t';
  }
  if (isString && isPrintable(r)) {
    return String.fromCodePoint(r);
  }
  if (r >= 0x20 && r <= 0x7E) {
    return String.fromCharCode(r);
  }
  if (r < 0x100) {
    return '\\x' + r.toString(16).padStart(2, '0');
  }
  if (r > 0xFFFF) {
    return '\\U' + r.toString(16).padStart(8, '0');
  }
  return '\\u' + r.toString(16).padStart(4, '0');
}

/**
 * @param {string} s
 * @param {LiteralQuoteConfig} cfg
 */
export function toStringLiteral(s, cfg) {
  const quote = quoteCharForPayload(s, cfg);
  const parts = [quote];
  for (const ch of s) {
    parts.push(escapeRune(ch.codePointAt(0) ?? 0, true, quote));
  }
  parts.push(quote);
  return parts.join('');
}

/**
 * @param {Uint8Array} data
 * @param {LiteralQuoteConfig} cfg
 */
export function toBytesLiteral(data, cfg) {
  const quote = quoteCharForPayload(data, cfg);
  const parts = ['b', quote];
  for (const b of data) {
    parts.push(escapeRune(b, false, quote));
  }
  parts.push(quote);
  return parts.join('');
}

/**
 * @param {string} payload
 * @param {string} castType
 * @param {LiteralQuoteConfig} cfg
 */
export function sqlCastQuoted(payload, castType, cfg) {
  const lit = toStringLiteral(payload, cfg);
  return `CAST(${lit} AS ${castType})`;
}

File: rust/src/quote.rs
//! GoogleSQL literal quoting and escaping.


#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum QuoteStrategy {
    Legacy = 0,
    Always = 1,
    MinEscape = 2,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum PreferredQuote {
    Double = 0,
    Single = 1,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct LiteralQuoteConfig {
    pub strategy: QuoteStrategy,
    pub preferred_quote: PreferredQuote,
}

impl Default for LiteralQuoteConfig {
    fn default() -> Self {
        Self {
            strategy: QuoteStrategy::Legacy,
            preferred_quote: PreferredQuote::Double,
        }
    }
}

pub fn normalize_literal_quote(cfg: LiteralQuoteConfig) -> LiteralQuoteConfig {
    let strategy = match cfg.strategy {
        QuoteStrategy::Legacy | QuoteStrategy::Always | QuoteStrategy::MinEscape => cfg.strategy,
    };
    let preferred_quote = match cfg.preferred_quote {
        PreferredQuote::Double | PreferredQuote::Single => cfg.preferred_quote,
    };
    LiteralQuoteConfig {
        strategy,
        preferred_quote,
    }
}

pub fn quote_char_for_payload(payload: &[u8], cfg: LiteralQuoteConfig) -> char {
    let cfg = normalize_literal_quote(cfg);
    if cfg.strategy == QuoteStrategy::Always {
        return if cfg.preferred_quote == PreferredQuote::Single {
            '\''
        } else {
            '"'
        };
    }

    let pref = if cfg.preferred_quote == PreferredQuote::Single {
        b'\''
    } else {
        b'"'
    };
    let other = if pref == b'\'' { b'"' } else { b'\'' };

    if cfg.strategy == QuoteStrategy::MinEscape {
        let single_count = payload.iter().filter(|&&b| b == b'\'').count();
        let double_count = payload.iter().filter(|&&b| b == b'"').count();
        if single_count < double_count {
            return '\'';
        }
        if double_count < single_count {
            return '"';
        }
        return pref as char;
    }

    let mut has_pref = false;
    for &b in payload {
        if b == other {
            return pref as char;
        }
        if b == pref {
            has_pref = true;
        }
    }
    if has_pref {
        other as char
    } else {
        pref as char
    }
}

fn is_printable(r: u32) -> bool {
    static BITSET: &[u8] = include_bytes!("unicode_printable.bin");
    if r >= 0x110000 {
        return false;
    }
    let idx = (r / 8) as usize;
    let bit = 1u8 << (r % 8);
    BITSET[idx] & bit != 0
}

pub fn escape_rune(r: u32, is_string: bool, quote: Option<char>) -> String {
    let q = quote.map(|c| c as u32).unwrap_or(u32::MAX);
    if r == b'\\' as u32 || (quote.is_some() && r == q) {
        return format!("\\{}", char::from_u32(r).unwrap_or('\u{fffd}'));
    }
    if is_string && r == b'\n' as u32 {
        return "\\n".to_string();
    }
    if is_string && r == b'\r' as u32 {
        return "\\r".to_string();
    }
    if is_string && r == b'\t' as u32 {
        return "\\t".to_string();
    }
    if is_string && is_printable(r) {
        return char::from_u32(r).unwrap_or('\u{fffd}').to_string();
    }
    if (0x20..=0x7E).contains(&r) {
        return char::from_u32(r).unwrap_or('\u{fffd}').to_string();
    }
    if r < 0x100 {
        return format!("\\x{r:02x}");
    }
    if r > 0xFFFF {
        return format!("\\U{r:08x}");
    }
    format!("\\u{r:04x}")
}

pub fn to_string_literal(s: &str, cfg: LiteralQuoteConfig) -> String {
    let quote = quote_char_for_payload(s.as_bytes(), cfg);
    let mut out = String::with_capacity(s.len() + 2);
    out.push(quote);
    for ch in s.chars() {
        out.push_str(&escape_rune(ch as u32, true, Some(quote)));
    }
    out.push(quote);
    out
}

pub fn to_bytes_literal(data: &[u8], cfg: LiteralQuoteConfig) -> String {
    let quote = quote_char_for_payload(data, cfg);
    let mut out = String::with_capacity(data.len() + 3);
    out.push('b');
    out.push(quote);
    for &b in data {
        out.push_str(&escape_rune(u32::from(b), false, Some(quote)));
    }
    out.push(quote);
    out
}

pub fn sql_cast_quoted(payload: &str, cast_type: &str, cfg: LiteralQuoteConfig) -> String {
    let lit = to_string_literal(payload, cfg);
    format!("CAST({lit} AS {cast_type})")
}
